src/coreclr/jit/emit.cpp

```diff
@@ -4479,6 +4610,429 @@ void emitter::emitJumpDistBind()
+{
+    if ((emitComp->opts.compJitAlignLoopBoundary > 16) && (!emitComp->opts.compJitAlignLoopAdaptive))
+    {
+        emitLongLoopAlign(emitComp->opts.compJitAlignLoopBoundary);
+    }
+    else
+    {
+        emitLoopAlign();
+    }
+
+    // Mark this IG as need alignment so during emitter we can check the instruction count heuristics of
+    // all IGs that follows this IG and participate in a loop.
+    emitCurIG->igFlags |= IGF_LOOP_ALIGN;
+
+    JITDUMP("Adding 'align' instruction of %d bytes in G_M%03u_IG%02u.\n", emitComp->opts.compJitAlignLoopBoundary,
+            emitComp->compMethodID, emitCurIG->igNum);
+}
+
+//-----------------------------------------------------------------------------
+//  For loopHeaderIg, find the size of the smallest possible loop that doesn't exceed maxLoopSize.
+//
+unsigned emitter::getLoopSize(insGroup* igLoopHeader, unsigned maxLoopSize)
+{
+    unsigned loopSize = 0;
+
+    for (insGroup* igInLoop = igLoopHeader; igInLoop != nullptr; igInLoop = igInLoop->igNext)
+    {
+        loopSize += igInLoop->igSize;
+        if (igInLoop->isLoopAlign())
+        {
+            // If igInLoop's next IG is a loop and needs alignment, then igInLoop should be the last IG
+            // of the current loop and should have backedge to current loop header.
+            assert(igInLoop->igLoopBackEdge == igLoopHeader);
+
+            // In such cases, the current loop size should exclude the align instruction size reserved for
+            // next loop.
+            unsigned maxPaddingAllowed;
+            if (emitComp->opts.compJitAlignLoopAdaptive)
+            {
+                maxPaddingAllowed = (emitComp->opts.compJitAlignLoopBoundary >> 1) - 1;
+            }
+            else
+            {
+                maxPaddingAllowed = emitComp->opts.compJitAlignLoopBoundary - 1;
+            }
+
+            loopSize -= maxPaddingAllowed;
+        }
+        if ((igInLoop->igLoopBackEdge == igLoopHeader) || (loopSize > maxLoopSize))
+        {
+            break;
+        }
+    }
+
+    return loopSize;
+}
+
+//-----------------------------------------------------------------------------
+// emitSetLoopBackEdge : Sets igLoopBackEdge field, if not already set and
+//                       if currIG has back-edge to dstIG.
+//
+// Notes:
+//    If the current loop covers a loop that is already marked as align, then remove
+//    the alignment flag present on IG before dstIG.
+//
+void emitter::emitSetLoopBackEdge(BasicBlock* loopTopBlock)
+{
+    insGroup* dstIG = (insGroup*)loopTopBlock->bbEmitCookie;
+
+    // With (dstIG != nullptr), ensure that only back edges are tracked.
+    // If there is forward jump, dstIG is not yet generated.
+    //
+    // We don't rely on (block->bbJumpDest->bbNum <= block->bbNum) because the basic
+    // block numbering is not guaranteed to be sequential.
+
+    if ((dstIG != nullptr) && (dstIG->igNum <= emitCurIG->igNum) && (emitCurIG->igLoopBackEdge == nullptr))
```

Comment:
Seems odd to me that we track the "first" backedge and not the "last". But perhaps we don't align loops with multiple back-edges today...?

Reply:
I am ok one way or the other ("first" vs. "last"). My thinking behind tracking "first" was that as long as there is one path that completes a loop, use its size to determine if we should align or not. On the contrary, we can argue that it might not be best possible way to track the smallest size and we should take into consideration the maximum length of the loop. I don't have enough data to prove which one is better.

> But perhaps we don't align loops with multiple back-edges today...?

I don't think it will matter if we choose "first" or "last". With both cases, we can align loops with multiple back-edges without needing to have special handling, unless I am missing something here.

Reply:
I have change this to `"last"` back-edge. In .NET libraries, I do not see any difference in no. of loops aligned but as per offline conversation, this "last" is better to do.